testing/tests/DevExpress.ui.widgets.dataGrid/editing.tests.js

```diff
@@ -3200,6 +3200,83 @@ QUnit.test('Remove the inserted row with edit mode batch and hidden column', fun
     assert.ok(!testElement.find('tbody > tr').first().hasClass('dx-row-inserted'), 'not has row inserted');
 });
 
+QUnit.test('AddRow method should return Deferred (T844118)', function(assert) {
+    // arrange
+    const done = assert.async();
+
+    this.options.editing = {
+        mode: 'batch',
+        allowAdding: true
+    };
+
+    this.rowsView.render($('#container'));
+
+    // assert
+    assert.equal(this.getVisibleRows().length, 7, '7 visible rows');
+
+    // act
+    const deferred = this.addRow();
+    deferred.done(() => {
+        // assert
+        assert.ok(true, 'result is Deferred');
+        assert.equal(this.getVisibleRows().length, 8, 'one more row is added');
+
+        done();
+    });
+});
+
+QUnit.test('Sequential adding of a row after adding the previous using Deferred (T844118)', function(assert) {
+    // arrange
+    const that = this;
+    const done = assert.async();
+    let promiseInitNewRowCallCount = 0;
+    let initNewRowCallCount = 0;
+
+    that.options.editing = {
+        mode: 'cell',
+        allowAdding: true
+    };
+
+    that.options.onInitNewRow = (e) => {
+        if((promiseInitNewRowCallCount + initNewRowCallCount) % 2 === 0) {
```

Comment:
Why does such complex condition use?

Reply:
You can define deferred outside onInitNewRow and resolve it manually if you want test that addRow promise is resolved after resolving e.promise in onInitNewRow

Reply:
Done